### ISSUE TITLE
Any chance of ".Net Core" for cross-platform?

### ISSUE THREAD

**Issue by felaugmar** [MonoGame/MonoGame#5339]
Hey, for now, Monogame's using [Mono](http://www.mono-project.com/) for targeting multiple platform.
My question is: Could [.NET Core](https://www.microsoft.com/net/core/platform) be used instead of [Mono](http://www.mono-project.com/)?

**Comment by DDReaper** [comment]
A full .NET Core / Standard library wouldn't be possible as it doesn't yet support all the necessary API's for game development, such as Input, Graphics and audio, to name but a few.

However it may be a more compatible and viable alternative to the .Core solution that was being proposed a time back, given that .NET core projects are even more accessible that PCL's it may be more useful.

So @tomspilman @dellis1972 @KonajuGames What are your thoughts?  Is there still an appetite to move all the platform generic code in to a central library in a future release (obviously not 3.6) and if so, does a .NET Core project sounds like a better alternative to a PCL.

**Comment by cra0zy** [comment]
.Net core is still not quite there yet, few of the major problems that it has are:
 - No support for .config files so all the wrapper files would need to be compiled once for each system
 - No support for a lot of Linux distros
 - Most IDEs still don't support .csproj .Net Core projects
 - AFAIK Protobuild doesn't support generating .Net Core projects

> Hey, for now, Monogame's using Mono for targeting multiple platform.
> My question is: Could .NET Core be used instead of Mono?

I have seen so many of these questions, so my question for you is why would you want .Net Core support?

> A full .NET Core / Standard library wouldn't be possible as it doesn't yet support all the necessary API's for game development, such as Input, Graphics and audio, to name but a few.

Mono doesn't offer any of that and we use it... that is not the problem with .Net Core.

> However it may be a more compatible and viable alternative to the .Core solution that was being proposed a time back, given that .NET core projects are even more accessible that PCL's it may be more useful.

Comparing PCL and .Net Core projects is like comparing PCL and C# Library projects.

**Comment by gulbanana** [comment]
One advantage of .net core is that it's easy to embed a standalone clr/corelibs with a packaged app.

**Comment by cra0zy** [comment]
> One advantage of .net core is that it's easy to embed a standalone clr/corelibs with a packaged app.

You can do the same thing with mono, just use mkbundle.

**Comment by gulbanana** [comment]
That's good to know, previously I'd only come across the kickstart.

**Comment by aienabled** [comment]
Interesting read:
Building a 3D Game Engine with .NET Core
https://mellinoe.wordpress.com/2017/01/18/net-core-game-engine/

**Comment by RUSshy** [comment]
Performance wise Mono is far behind, .NET core is the future, and is set to completly replace Mono

Would be cool to see more support for it

**Comment by cra0zy** [comment]
All the problems I described above are still present :/

**Comment by cra0zy** [comment]
Well I got MonoGame working with .Net Core today:
![dotnetcore](https://cloud.githubusercontent.com/assets/6773302/23760029/b63be1e6-04ee-11e7-964b-32c567d7e86c.png)

Don't expect to see this in the repo until the problems I described get addressed.

**Comment by DDReaper** [comment]
Fantastic news @cra0zy Although can you detail a little more what you mean by "working on .Net core"?
So is it a "core" version of MG, a Netcore project referencing MG as is, etc?

**Comment by cra0zy** [comment]
I ported MonoGame DesktopGL to use `netstandard1.4`, it's pretty much the full version of MonoGame (except Song stuff, I had to cut that part out, see #5562).

**Comment by DDReaper** [comment]
An interesting start @cra0zy , would be interesting to see that also compiled against netstandard2.0 to see what other benefits it might garner.
Do you have a branch worth checking or is al local work?

**Comment by cra0zy** [comment]
No branch, all local, I can send you the files, but it's nothing special. I do plan on getting some stuff fixed in the main repo so it's easier to integrate .net core later on.

**Comment by dsaf** [comment]
Isn't Xamarin going to properly support .NET Standard only starting with 2.0?

PS: thanks for the MonoGame PCL 3.6 - it's amazing.

**Comment by Jjagg** [comment]
I can find awfully little information about timeframes. Maybe @dellis1972 knows something we don't?

**Comment by cra0zy** [comment]
I took a bit of time to fix up sound and upload MG .net core version if someone wants to play around with: https://github.com/cra0zy/MonoGameCore

**Comment by dsaf** [comment]
This might allow using C# 7 it seems https://github.com/dotnet/corefx/issues/14235

Roadmap for .NET standard: https://github.com/dotnet/core/blob/master/roadmap.md

**Comment by Genbox** [comment]
Right now, it is pretty difficult to implement MonoGame in a cross-platform project that uses .NET core, as you have to target net45 everywhere to use it.

**Comment by dsaf** [comment]
@Genbox I am extremely glad to see Farseer coming back to life!

**Comment by cra0zy** [comment]
The current major problem with MG supporting .net core is: https://github.com/dotnet/coreclr/issues/930

There are several ways we could workaround it, but I would rather see it properly implemented in .net core.

**Comment by Genbox** [comment]
It took me 47 minutes to read all of that. Phew... that was not pleasant.
In a month, the issue will be 2 years old. I don't think it makes sense to hold our breath until an official solution for DLL mapping exist. 

It seems that AssemblyLoadContext (part of System.Runtime.Loader) is capable of doing what you need. Create your own assembly loader that inherits from AssemblyLoadContext and override LoadUnmanagedDll. Then you have to call LoadLibrary (or similar) to obtain a handle, which you return. Since .NET Core uses AssemblyLoadContext.Default and not your assembly loader, subscribe to the AssemblyLoadContext.Default.Resolving event, which you can then pass your assembly loader on to.

[DLLImport("mylib")] would then result in a call to Resolving, in which you provide your assembly load context, which then maps "mylib" to the right DLL for the platform reported by the RuntimeInformation class.

It is a little more legwork than a simple .config file with the mapping, but it seems trivial to implement. @cra0zy what are your thoughts on this?

**Comment by cra0zy** [comment]
> @cra0zy what are your thoughts on this?

I currently know 5 ways of "solving" this and the way I would prefer to do it is by overriding library folder and having a constant nameset for all the libraries. Alternative easy solution is just spiting DesktopGL into 3 assemblies (one for each platform) and `#ifdef`-ing the const string for the lib name.

**Comment by cra0zy** [comment]
Side note, the 5 ways are:
 - Modify library load folder, set all library names to same name
 - Split DesktopGL into 3 assemblies
 - Nightmare you described :P
 - T4Template to generate 3 sets of functions
 - Separate assembly which contains the constant and is added as a dependency

**Comment by zigzag312** [comment]
I don't know if this is of any use here. There is a hack on Windows to load DLL with function `LoadLibrary` in static constructor or at the beginning of the program and `DllImport` will then use cached DLL that you already loaded manually. [Example1](http://stackoverflow.com/questions/10852634/using-a-32bit-or-64bit-dll-in-c-sharp-dllimport/30646096#30646096), [example2](http://stackoverflow.com/questions/11934570/change-c-sharp-dllimport-target-code-depending-on-x64-x86/11936113#11936113)

Maybe this hack could be modified somehow to work on other platforms too?

**Comment by Genbox** [comment]
@cra0zy It certainly seems doable. I'd say the changes that are needed are small enough that it can be re-done if the .NET Core team ever decides on this issue.

**Comment by cra0zy** [comment]
@Genbox All the 5 ways I described make at least 1 new problem :|

**Comment by Genbox** [comment]
@cra0zy I certainly don't have the same insight into MonoGame as you have, so I defer that to you. However, working towards the .NET Core platform and being part of the .NET Standard is a big deal and will solve many problems that library authors have with targeting to different platforms.

**Comment by zigzag312** [comment]
Portable Class Libraries (PCLs) now officially deprecated [16 August 2017]

> If you’re sharing code between different .NET implementations today, you’re probably aware of Portable Class Libraries (PCLs). With the release of .NET Standard 2.0, we’re now officially deprecating PCLs and you should move your projects to .NET Standard.

Source: Announcing .NET Standard 2.0

**Comment by tomspilman** [comment]
https://blogs.msdn.microsoft.com/dotnet/2017/08/14/announcing-net-standard-2-0/

@DDReaper @dellis1972 - So what does this mean for us?

The article says UWP doesn't support .NET Standard 2.0... we we should wait to move over to it at least until then.  Also i want to be sure that we can support it on the console builds too.

Still we should do all the prep work we can to get MG ready to support it.

I suspect our existing bait-and-switch PCL library will still work like it does, but it just will use .NET Standard 2.0 and not the weird PCL profiles it does right now.

This make sense?

**Comment by cra0zy** [comment]
I think we will first need .Net Core versions of WindowsDX and DesktopGL versions so we can properly convert it to a PCL like .net core version.

Protobuild still doesn't support .Net Core btw.

**Comment by tomspilman** [comment]
@hach-que - I know you were thinking about .NET Core/Standard a few years ago.  It is starting to look like we can really switch to it soon (like within 6 months).  What is left to do to support it in Protobuild?  Is this something you plan to work on?

**Comment by hach-que** [comment]
It's something I want to support, but we're super busy internally at the moment with the upcoming launch of [HiveMP](https://hivemp.com/). I'm probably several months away from really having time to spend on open source stuff again.

All that's needed in Protobuild is to update the XSLT files to support generating the C# project files. This could be done independently if you need it sooner than I can ship, by overriding the XSLT files in the MonoGame repo locally.

**Comment by zigzag312** [comment]
@tomspilman  UWP support was just added [10. October 2017]
But there are some limits to use .NET Standard 2.0 in UWP:

> You need to download Visual Studio 2017 15.4, also released today, to take advantage of these updates. You also need to be running on and targeting Windows 10 Fall Creators Update, which will be released later this month. .NET Standard 2.0 support for UWP required changes in Windows 10 that are shipping in Fall Creators Update. These changes established a new baseline for UWP tools.

[Announcing UWP Support for .NET Standard 2.0 | .NET Blog](https://blogs.msdn.microsoft.com/dotnet/2017/10/10/announcing-uwp-support-for-net-standard-2-0/)

@cra0zy .NET Standard 2.0 adds many missing APIs (reflection APIs, non-generic collections, DataSet, binary serialization, XML Schema etc.) so supporting .NET Core should be easier if we can target .NET Standard 2.0?

I would like to point out that .NET Core & .NET Standard [are not the same thing](https://stackoverflow.com/questions/42939454/what-is-the-difference-between-net-core-and-net-standard-class-library-project?rq=1). .NET Core works only on new CoreCLR runtime while .NET Standard is a replacement for PCL and supports .NET Framework, Mono, Xamarin, UWP & .NET Core. So if we create .NET Standard version .NET Core is automatically supported.

Regarding Protobuild, I may be wrong as I don't have much experience with it, but do we still need Protobuild with the new .csproj format? 
The new .csproj format is greatly simplified. It doesn't need list of all .cs files included in project as it includes all *.cs files by default. I think we could just create separate project files and add globs to .csproj to exclude files/folders that do not belong to specific platform (for example exclude all *.DirectX.cs files).

**Comment by cra0zy** [comment]
> @cra0zy .NET Standard 2.0 adds many missing APIs (reflection APIs, non-generic collections, DataSet, binary serialization, XML Schema etc.) so supporting .NET Core should be easier if we can target .NET Standard 2.0?

It won't be hard either way, when I was doing the port .NET Standard 2.0 didn't exists, and I still managed to do it with minimal changes: https://github.com/cra0zy/MonoGameCore/blob/master/MonoGame.Framework/MonoGame.Framework.csproj#L4

> I would like to point out that .NET Core & .NET Standard are not the same thing.

I didn't realize there was a .NET Core Library... I'll be less ambiguous from now on when I talk about this.

> Regarding Protobuild, I may be wrong as I don't have much experience with it, but do we still need Protobuild with the new .csproj format?

Yup, I think so.

**Comment by cra0zy** [comment]
Well I decided to update the repo (well in this case move stuff over to fork), upgrade the required .net core version to 2.0 and setup some NuGet packages for templates.

**TLDR:**
```sh
dotnet new --install MonoGame.Template.CSharp

# in directory where you want to create the project
dotnet new mgdesktopgl
dotnet run
```

**Long version:**
So since this isn't gonna come to the main repo any time soon, and I have gave up all hope on MonoDevelop I wanted to setup some MG DesktopGL templates for myself on Linux. Therefore I have moved the repo to the fork: https://github.com/cra0zy/MonoGame/tree/core and added all the new shiny stuff that was required to setup some nuget packages for it.

Now the list of needed packages for this are:
- https://www.nuget.org/packages/MonoGame.Content.Builder/
- https://www.nuget.org/packages/MonoGame.Framework.DesktopGL.Core/
- https://www.nuget.org/packages/MonoGame.Template.CSharp/

Do note that Content.Builder nuget is ~120MB in size, tho once it is downloaded by .net core it goes into the local nuget repo so it won't be downloaded again.

Now its already 3 in the morning in my time zone and I really need to sleep so if you find some bugs, post them bellow and I'll see if I can resolve them tomorrow. Good night world.

PS. Rider picks up all .Net core templates so in some strange way this is also a Rider IDE Template for MG :)

PPS. @hach-que @zigzag312 https://github.com/cra0zy/MonoGame/blob/core/MonoGame.Framework/MonoGame.Framework.DesktopGL.Core.csproj#L3

**Comment by hach-que** [comment]
Thanks for the reference. Feel free to send a PR to Protobuild if you want .NET Core support (it doesn't look like too many changes are needed). I'm flat out with game dev stuff so unless there's something critical that comes up for MonoGame, it's unlikely I'll get a chance to dev for it.

**Comment by zigzag312** [comment]
@cra0zy are you referring to csproj file or EnableDefaultCompileItems element? What about it?

**Comment by cra0zy** [comment]
@zigzag312 the element, gets the old csproj behavior.

**Comment by jkotas** [comment]
> having a constant nameset for all the libraries

It is the way to do this in .NET Core. The build tooling and `dotnet` runner have built-in support for this. You just need to put the native libraries into platform specific folders in the NuGet package and the rest will happen automatically. You do not need to override any paths yourself.

Here is an example you can see it in action. Microsoft.AspNetCore.Server.Kestrel.Transport.Libuv is a NuGet package with unmanaged .dll that is different for each platform.

```
dotnet new console
dotnet add package Microsoft.AspNetCore.Server.Kestrel.Transport.Libuv
dotnet publish
dir /s bin\Debug\netcoreapp2.0\publish\*libuv* // you should see all variants of libuv published
```

Also, there are several ways how to filter what gets published. For example, if you publish self-contained app using `dotnet publish -r win-x64`, you will only get the win-x64 version in the publish folder.

**Comment by cra0zy** [comment]
@jkotas The nuget above is already doing it, that problem has been fixed for quite some time now.

**Comment by Majirefy** [comment]
@cra0zy can you share some experience on how to create a .Net Core 2.0 MonoGame from scratch?

I created a .Net Core 2.0 console application on VS Mac, then which NuGet package should I install?
How to set up iOS and Android project?

**Comment by cra0zy** [comment]
https://github.com/MonoGame/MonoGame/issues/5339#issuecomment-342686703

**Comment by Trapov** [comment]
@cra0zy Will it be ever merged? Or is it going to be a fork?

Also what to do you think about distrubutuing a game-app which targets dotnetcore, what's your thoughts on it? Like, with mono we supposed to drag some parts of the mono-runtime with us, isn't it going to be much(file size) with dotnetcore?

**Comment by cra0zy** [comment]
> Will it be ever merged, or is it going to be a fork?

Well the plan is to get this merged in 3.8.

> And is it already duable to work with, or just rather use mono?

Not sure what you are asking here, but .NET Standard libraries can be used both from .NET Core apps and standard mono apps.

> Also what to do you think about distrubutuing a game-app which targets dotnetcore, what's your thoughts on it? With mono we supposed to drag some parts of the mono-runtime with us, isn't it going to be much(file size) with dotnetcore?

With .NET Core you also need to package up the runtime, however there is already a built in command `dotnet publish` to help you accomplish this task, Plus you can use ILLinker to further minimize your app size. Other than that there aren't that many differences between the two.

PS. For Windows, you could also try to compile your app without .NET runtime by using corert: https://github.com/dotnet/corert/tree/master/samples/MonoGame , its very experimental tho.

**Comment by Trapov** [comment]
> Not sure what you are asking here, but .NET Standard libraries can be used both from .NET Core apps and standard mono apps.

@cra0zy, i mean sure, i can use them anywhere i can, since .net standard unites the platforms pretty much. But i'm referring to your comments above:

> , it's pretty much the full version of MonoGame 

Does it lack many API's, or is it _the same_? 

My question is - will i lose something if i try to start doing my app with dotnetcore?

thanks for replies.

**Comment by cra0zy** [comment]
> Does it lack many API's, or is it the same?
> 
> My question is - will i lose something if i try to start doing my app with dotnetcore?
> 
> thanks for replies.

Its the full version, everything is there.

**Comment by tomspilman** [comment]
This is great...

https://github.com/dotnet/corert/tree/master/samples/MonoGame

**Comment by jvlppm** [comment]
I am trying to build my old project with the packages by @cra0zy
```xml
    <PackageReference Include="MonoGame.Content.Builder" Version="3.7.0.4" />
    <PackageReference Include="MonoGame.Framework.DesktopGL.Core" Version="3.7.0.7" />
```
and the template feels awesome,

but in my code I have a content pipeline extension and I can't find references to:

```csharp
using Microsoft.Xna.Framework.Content.Pipeline;
```
or the class `ContentImporter`

In the past I used the Portable Pipeline Nuget package (PCL)

Any chance of me getting this to work with .Net Standard?
How can I proceed?

I found this repo https://github.com/craftworkgames/MonoGame.Extended
maybe I just have to wait for their 2.0 version which is currently in development

**Comment by LithiumToast** [comment]
@jvlppm Extended is also having an issue with the content pipeline for .NET Core. You can read about it [here for a workaround hack](https://github.com/craftworkgames/MonoGame.Extended/issues/495).

**Comment by cra0zy** [comment]
> In the past I used the Portable Pipeline Nuget package (PCL)

You can (and should) use the Portable Pipeline Nuget still, this does not change anything about content pipeline extension compilation.

**Comment by craftworkgames** [comment]
I just wanted to say I really appreciate the work that's gone into this. I know it's a difficult problem to solve. We're currently migrating MonoGame.Extended over to .NET Standard. It's going to be quite a while before we release 2.0 but it feels like we're heading in the right direction.

**Comment by Thraka** [comment]
What about getting this to work with UWP?

**Comment by mrhelmut** [comment]
Just to let you know that we successfully ported multiple full games to .NET Core and built them with CoreRT based on the work of @cra0zy.

It works very well. Performance wise, games start very quickly but gameplay performance are the same (which was to be expected in an AOT context).
But the real advantages of CoreRT is that it allows to obfuscate the exe, and most importantly, to produce standalone exe with absolutely 0 dependencies. Everything is statically linked, including the .NET runtime, and any MVC runtime as well.
This is super cool for shipping desktop games that work as-is.

I have only tested CoreRT on Windows for now, but this is surely something to keep developing.

On the other hand, I tried to AOT compile games using Mono, but to no avail. I'm still getting errors on native libraries marshaling producing IL code, but the main drawback of Mono is that even if assemblies are AOT compiled, Mono still requires the original IL assemblies to be present, which prevents the possibility to obfuscate them.

**Comment by Jjagg** [comment]
@mrhelmut That's awesome :) What about filesize? I found including the .NET runtime bloated the filesize by quite a bit. Any plansto write more about the process? (please take that as a question, not a request; I know you're busy)

**Comment by mrhelmut** [comment]
File size gets bloated, but not that much, about 10-15MB. That's very little compared to what BRUTE produce or even embedding Mono with MonoKickstart (which is the biggest overhead, ~55MB on Linux).

There's no benefit to using Mono, expect that it would allow to AOT compile games as a post-process. (Now that you mention it, I believe that I know why my AOT attempt failed: I tried to mono AOT assemblies built with the .Net compiler. I believe that building with Mono behave differently regarding delegate marshaling.)

Regarding the process, it was as simple as switching to @cra0zy .Net Core's template and MG branch, and target CoreRT.

**Comment by Jjagg** [comment]
Neat :) Thanks for the quick response :+1:

**Comment by cra0zy** [comment]
You should be able to use ILLink.Tasks for extra size reduction.

**Comment by Jjagg** [comment]
Closing this in favor of #4975.

**Comment by cra0zy** [comment]
@Jjagg #4975 is asking about a different subject.

**Comment by Jjagg** [comment]
@cra0zy It's all about moving the lib to .NET Standard. Though #4975 specifically is about replacing PCL. Maybe better to open a new issue 'Support .NET Standard' to replace both this and #4975.

**Comment by TheFanatr** [comment]
Just FYI, it is possible to run the latest version of MonoGame for Windows 10 desktop (which has it's templates currently targeting .NET Framework) on the .NET Core 3.0 Preview 2 runtime, as long as the newly-preview-released `Microsoft.NET.Sdk.WindowsDesktop` SDK is targeted in the project. I have it running in Visual Studio 2019 with C# 8 and it works perfectly. Templates for that specific part could be made already. As for supporting other platforms, something else needs to be done, but to reiterate, this is the vanilla `MonoGame.Framework` running on .NET Core for desktop Windows.

**Comment by cra0zy** [comment]
Hmm... 🤔 , .NET Core is not supposed to be able to reference .NET Framework assemblies, so something is very fishy with that .csproj file. I'll wait till .NET Core 3 gets officially released to test stuff out.

**Comment by jkotas** [comment]
> .NET Core is not supposed to be able to reference .NET Framework assemblies

That is not correct. Read ".NET Framework compatibility mode" in https://blogs.msdn.microsoft.com/dotnet/2017/08/14/announcing-net-standard-2-0/

**Comment by TheFanatr** [comment]
Last time I checked, .NET Core was always able to use a compatibility shim to access assemblies targeting .NET Framework.

**Comment by mrhelmut** [comment]
It's a good hint toward the fact that it can run on .Net Core, but moving to .Net Standard still is the way to go to future-proof MonoGame and solve cross-platform deployment (as well as supporting CoreRT out of the box). Probably safe to wait for .Net Core 3.

**Comment by cra0zy** [comment]
> That is not correct. Read ".NET Framework compatibility mode" in https://blogs.msdn.microsoft.com/dotnet/2017/08/14/announcing-net-standard-2-0/

Ha, interesting, didn't know about that, thanks @jkotas

**Comment by cra0zy** [comment]
> as well as supporting CoreRT out of the box

I'm somewhat interested in knowing if the project above can be compiled with corert.

**Comment by RUSshy** [comment]
there is a monogame sample on corert repo https://github.com/dotnet/corert/tree/master/samples/MonoGame

**Comment by cra0zy** [comment]
> there is a monogame sample on corert repo https://github.com/dotnet/corert/tree/master/samples/MonoGame

I know, @jkotas added it, it uses my .NET Standard MG DesktopGL assemblies XD

The point of above question is mostly to know if .NET Framework assemblies can be compiled with corert.

**Comment by quixoticaxis** [comment]
[FYI DLLMap is coming to .Net Core](https://devblogs.microsoft.com/dotnet/announcing-net-core-3-preview-3/). I didn't dig in, but I suppose people will find it useful: inability to manage native dependencies in runtime was mentioned as one of the blockers for .Net Core migration.

**Comment by cra0zy** [comment]
Yea, MonoGame is already doing that except manually.

**Comment by mrhelmut** [comment]
@cra0zy @Jjagg Is there a place where to follow the effort and help moving forward? I feel like we really need to have a fully working .Net Core setup now (inc. templates, supporting mgcb, installing the pipeline from both source and nuget, removing the installer...). Can we put together a comprehensive to do list somewhere? I'd like to use the free time I have now to help.

**Comment by Jjagg** [comment]
We could publish the NuGets already, so maybe that's a good next step. I'll make a task list in #6768. Would be great if you can help, I don't have any spare time atm.

**Comment by mrhelmut** [comment]
Will check. I'm working on a bit to do issue to keep track of all the separated efforts.

**Comment by montyclt** [comment]
Any news now that .NET Core 3 is published?

**Comment by mrhelmut** [comment]
MonoGame is now fully compliant with .Net Core 3 and building from source will produce nugets that are consumable by ```dotnet``` apps. Several games are already using on net core 3 or even CoreRT.

There are still a few maintenance work to do on the repository (e.g. actually publishing those nugets and plugging them into the continuous integration; as well as making the documentation for them) before we can officially do a release.

**Comment by cra0zy** [comment]
Gonna close this as we have the todo list for the remaining stuff.
